File: packages/generator-core/src/buttons.ts
/** Form buttons: the actions a spec declares with root `buttons`, rendered in the form footer. */

import { FormInputError } from '@crudui/validator';

import { makeTranslate, type Language } from './content';
import { styleString } from './css';
import { resolveDesign } from './design';
import { makeContext } from './expr';
import type { FormTemplate } from './form';
import { formMessages } from './messages';
import { escAttr, escText } from './util';

/** Button kinds a spec can declare. A `link` renders an anchor. */
export type FormButtonType = 'submit' | 'reset' | 'button' | 'link';

/** Declared button types, in documentation order. */
export const FORM_BUTTON_TYPES: readonly FormButtonType[] = ['submit', 'reset', 'button', 'link'];

/** The buttons of a form whose spec declares none: one submit button. */
export const DEFAULT_FORM_BUTTONS: readonly Readonly<Record<string, unknown>>[] = [{ type: 'submit' }];

/** One evaluated form button. */
export interface ButtonVM {
  /** Declared button type. */
  type: FormButtonType;
  /** Element name: `a` for a link, `button` otherwise. */
  tag: 'a' | 'button';
  /** Declared content, or the interface text of the button type. */
  text: string;
  /** Attributes in output order: type, class, style, name, value, href, onclick. */
  attrs: Record<string, string>;
}

/** Options for evaluating form buttons. */
export interface BindButtonsOptions {
  /** Content and interface language, defaulting to Korean. */
  language?: Language;
}

/** A behavior script: a string, or the `script` of a `{ label, script }` action. */
function behaviorScript(behavior: unknown, action: string): string | undefined {
  const entry = behavior !== null && typeof behavior === 'object' ? (behavior as Record<string, unknown>)[action] : undefined;
  const script = entry !== null && typeof entry === 'object' ? (entry as Record<string, unknown>).script : entry;
  return typeof script === 'string' && script !== '' ? script : undefined;
}

/**
 * Evaluate the template's buttons for a record. Text is the declared content or the
 * interface text named by the button type; design classes and styles follow the
 * field design rules against the record data.
 */
export function bindButtons(
  template: FormTemplate,
  data: Record<string, unknown> = {},
  options: BindButtonsOptions = {},
): ButtonVM[] {
  if (template.kind !== 'crudui/form-template') throw new FormInputError('Unsupported form template');
  const language = options.language ?? 'ko';
  const t = makeTranslate(language);
  const messages = formMessages(language) as unknown as Readonly<Record<string, string | undefined>>;
  return template.buttons.map((spec): ButtonVM => {
    const type = spec.type as FormButtonType;
    const design = resolveDesign(spec.design, makeContext([], data));
    const style = styleString(design.main.style);
    const optional = { name: spec.name, value: spec.value, href: spec.href, onclick: behaviorScript(spec.behavior, 'onclick') };
    return {
      type,
      tag: type === 'link' ? 'a' : 'button',
      text: t(spec.text as never, messages[type]),
      attrs: {
        ...(type === 'link' ? {} : { type }),
        class: ['crudui-action', 'crudui-action--text', design.main.class].filter(Boolean).join(' '),
        ...(style ? { style } : {}),
        ...Object.fromEntries(Object.entries(optional).filter((entry): entry is [string, string] => typeof entry[1] === 'string')),
      },
    };
  });
}

/** The attributes a button may carry, in output order. */
const BUTTON_ATTRIBUTES = new Set(['type', 'class', 'style', 'name', 'value', 'href', 'onclick']);

const isObject = (value: unknown): value is Record<string, unknown> => value !== null && typeof value === 'object' && !Array.isArray(value);

/**
 * Markup of the form buttons. Every renderer inserts this one string into the footer controls
 * group. The buttons are evaluated buttons: a list of objects whose `tag` is `a` or `button`,
 * whose `text` is a string and whose `attrs` holds string values of the button attributes.
 */
export function formButtonsHtml(buttons: readonly ButtonVM[]): string {
  if (!Array.isArray(buttons)) throw new FormInputError('Form buttons must be a list');
  for (const button of buttons as unknown[]) {
    const evaluated = isObject(button) && (button.tag === 'a' || button.tag === 'button') && typeof button.text === 'string'
      && isObject(button.attrs) && Object.entries(button.attrs).every(([name, value]) => BUTTON_ATTRIBUTES.has(name) && typeof value === 'string');
    if (!evaluated) throw new FormInputError('Form buttons must be evaluated button objects');
  }
  return (buttons as readonly ButtonVM[]).map(button =>
    `<${button.tag}${Object.entries(button.attrs).map(([name, value]) => ` ${name}="${escAttr(value)}"`).join('')}>` +
    `${escText(button.text)}</${button.tag}>`).join('');
}

File: packages/generator-go/buttons.go
package generator

import (
	"fmt"
	"strings"
)

// formButtonTypes lists the button types a spec can declare. A link renders an anchor.
var formButtonTypes = map[string]bool{"submit": true, "reset": true, "button": true, "link": true}

// buttonText returns the interface text of a button type; only submit and reset have one.
func buttonText(m formMessages, kind string) string {
	switch kind {
	case "submit":
		return m.submit
	case "reset":
		return m.reset
	}
	return ""
}

// buttonLabel returns the declared content, or the interface text when there is none.
func buttonLabel(button *Object, language, fallback string) string {
	text := read(button, "text")
	if s, ok := text.(string); ok {
		return s
	}
	if object(text) != nil {
		if s := translate(text, language); s != "" {
			return s
		}
	}
	return fallback
}

// buttonScript returns a non-empty behavior script: a string or the script of a { label, script } action.
func buttonScript(button *Object, action string) string {
	v := read(read(button, "behavior"), action)
	if object(v) != nil {
		v = read(v, "script")
	}
	s, _ := v.(string)
	return s
}

// BindButtons evaluates the template's buttons for a record in declaration order.
// Each button is an object with type, tag, text and attrs; attrs keeps the output order
// type, class, style, name, value, href, onclick. Only the Language option applies.
func BindButtons(template *FormTemplate, data *Object, options BindOptions) ([]*Object, error) {
	if e := checkOrderedValue(data); e != nil {
		return nil, e
	}
	if template == nil || template.Kind != "crudui/form-template" {
		return nil, fmt.Errorf("Unsupported form template")
	}
	if data == nil {
		data = NewObject()
	}
	language, e := bindLanguage(options)
	if e != nil {
		return nil, e
	}
	m, e := messagesFor(language)
	if e != nil {
		return nil, e
	}
	return bindButtons(template.Buttons, data, language, m), nil
}

// bindLanguage returns the Language option, defaulting to ko and rejecting a non-string value.
func bindLanguage(options BindOptions) (string, error) {
	if options.Language == nil {
		return "ko", nil
	}
	s, ok := options.Language.(string)
	if !ok {
		return "", fmt.Errorf("Language must be a string")
	}
	return s, nil
}

// bindButtons evaluates declared buttons for a record; design rules read the record data.
func bindButtons(buttons []*Object, data *Object, language string, m formMessages) []*Object {
	lookup, _ := plainLookup(data).(map[string]any)
	out := make([]*Object, 0, len(buttons))
	for _, button := range buttons {
		kind := stringAt(button, "type")
		design := resolveDesign(read(button, "design"), lookup, []string{})
		tag, attrs := "button", NewObject()
		if kind == "link" {
			tag = "a"
		} else {
			attrs.Set("type", kind)
		}
		class := "crudui-action crudui-action--text"
		if c := nodeClass(design, "main"); c != "" {
			class += " " + c
		}
		attrs.Set("class", class)
		if style := styleString(nodeStyle(design, "main")); style != "" {
			attrs.Set("style", style)
		}
		for _, name := range []string{"name", "value", "href"} {
			if s, ok := read(button, name).(string); ok {
				attrs.Set(name, s)
			}
		}
		if script := buttonScript(button, "onclick"); script != "" {
			attrs.Set("onclick", script)
		}
		out = append(out, NewObject("type", kind, "tag", tag, "text", buttonLabel(button, language, buttonText(m, kind)), "attrs", attrs))
	}
	return out
}

// buttonAttributes lists the attribute names an evaluated button may carry.
var buttonAttributes = map[string]bool{"type": true, "class": true, "style": true, "name": true, "value": true, "href": true, "onclick": true}

// FormButtonsHTML renders evaluated buttons as the markup every renderer inserts into the form footer.
// Each button must be an object whose tag is a or button, whose text is a string and whose attrs
// object has only the named button attributes with string values; other members are ignored.
func FormButtonsHTML(buttons []*Object) (string, error) {
	var out strings.Builder
	for _, button := range buttons {
		tag, _ := read(button, "tag").(string)
		text, textOK := read(button, "text").(string)
		attrs := object(read(button, "attrs"))
		if button == nil || (tag != "a" && tag != "button") || !textOK || attrs == nil {
			return "", errNotEvaluatedButtons
		}
		out.WriteString("<" + tag)
		for _, name := range attrs.Keys() {
			value, ok := read(attrs, name).(string)
			if !ok || !buttonAttributes[name] {
				return "", errNotEvaluatedButtons
			}
			out.WriteString(" " + name + `="` + rawEscape(value) + `"`)
		}
		out.WriteString(">" + escapeText(text) + "</" + tag + ">")
	}
	return out.String(), nil
}

// errNotEvaluatedButtons rejects form button input that bindButtons did not produce.
var errNotEvaluatedButtons = fmt.Errorf("Form buttons must be evaluated button objects")
